File: src/storeFx/fetchOneArtist.ts
import * as Fx from '../fx/fx';
import { mergeIntoMap } from '../mergeIntoMap/mergeIntoMap';
import type { Store } from '../store/types';
import { getArtist } from '../subsonic';

export function make(artistId: string) {
  return Fx.bind(Fx.ask<{ store: Store }>(), ({ store }) =>
    Fx.bind(
      Fx.provide(getArtist(artistId), {
        credentials: store.getState().auth.credentials,
      }),
      ({
        album = [],
        albumCount,
        coverArt,
        musicBrainzId,
        name,
        starred,
        userRating,
        id,
      }) => {
        const state = store.getState();

        const albumBaseById = mergeIntoMap(
          state.albums.baseById,
          album,
          x => x.id,
        );

        const albumIdsByArtistId = mergeIntoMap(
          state.artists.albumIdsByArtistId,
          [album.map(x => x.id)],
          () => id,
        );

        const byId = mergeIntoMap(
          state.artists.byId,
          [
            {
              albumCount,
              coverArt,
              id,
              musicBrainzId,
              name,
              starred,
              userRating,
            },
          ],
          x => x.id,
        );

        if (
          albumBaseById !== state.albums.baseById ||
          albumIdsByArtistId !== state.artists.albumIdsByArtistId ||
          byId !== state.artists.byId
        ) {
          store.setState(prevState => ({
            albums: {
              ...prevState.albums,
              baseById: albumBaseById,
            },
            artists: {
              ...prevState.artists,
              albumIdsByArtistId,
              byId,
            },
          }));
        }

        return Fx.Ok();
      },
    ),
  );
}

File: src/storeFx/fetchOneArtist.ml
open Fx
open MergeIntoMap
open Monad_syntax
open Zustand

let make artistId =
  let* Deps.{ store } = ask () in
  let* response =
    let state = getState store in
    Subsonic.GetArtist.make artistId
    |. provide ~deps:{ credentials = state.auth.credentials }
  in
  let artist =
    Subsonic.BaseArtist.
      {
        albumCount = response.albumCount;
        coverArt = response.coverArt;
        id = response.id;
        musicBrainzId = response.musicBrainzId;
        name = response.name;
        starred = response.starred;
        userRating = response.userRating;
      }
  and state = getState store in
  let albums = response.album |> Option.value ~default:[||] in
  let albumBaseById = mergeIntoMap state.albums.baseById albums (fun x -> x.id)
  and albumIdsByArtistId =
    mergeIntoMap state.artists.albumIdsByArtistId
      [| Js.Array.map albums ~f:Subsonic.BaseAlbum.(fun x -> x.id) |]
      (fun _ -> response.id)
  and byId = mergeIntoMap state.artists.byId [| artist |] (fun x -> x.id) in
  if
    albumBaseById != state.albums.baseById
    || albumIdsByArtistId != state.artists.albumIdsByArtistId
    || byId != state.artists.byId
  then
    store
    |. setState (fun prevState ->
           {
             prevState with
             albums = { prevState.albums with baseById = albumBaseById };
             artists = { prevState.artists with albumIdsByArtistId; byId };
           });
  Ok ()
